modules/nf-core/limma/differential/templates/limma_de.R

```diff
@@ -255,84 +262,173 @@ if ((! is.null(opt\$exclude_samples_col)) && (! is.null(opt\$exclude_samples_val
+colnames(design) <- make.names(colnames(design))
+
+
+# Perform voom normalisation for RNA-seq data
+if (!is.null(opt\$use_voom) && opt\$use_voom) {
+    # Create a DGEList object for RNA-seq data
+    dge <- DGEList(counts = intensities.table)
+
+    # Normalize counts using TMM
+    dge <- calcNormFactors(dge, method = "TMM")
+
+    # Run voom to transform the data
+    voom_result <- voom(dge, design)
+    data_for_fit <- voom_result
+
+    # Write the normalized counts matrix to a TSV file
+    normalized_counts <- voom_result\$E
+    normalized_counts_with_genes <- data.frame(Gene = rownames(normalized_counts), normalized_counts, row.names = NULL)
+    colnames(normalized_counts_with_genes)[1] <- opt\$probe_id_col
+    write.table(normalized_counts_with_genes,
+        file = paste(opt\$output_prefix, "normalised_counts.tsv", sep = '.'),
+        sep = "\t",
+        quote = FALSE,
+        row.names = FALSE)
+
+} else {
+    # Use as.matrix for regular microarray analysis
+    data_for_fit <- as.matrix(intensities.table)
+}
+
 # Prepare for and run lmFit()
 
-lmfit_args = c(
-    list(
-        object = as.matrix(intensities.table),
-        design = design
-    ),
-    opt[c('ndups', 'spacing', 'block', 'method')]
+lmfit_args <- list(
+    object = data_for_fit,
+    design = design
 )
 
+# Include optional parameters if provided
+if (! is.null(opt\$ndups)){
+    lmfit_args[['ndups']] <- as.numeric(opt\$ndups)
+}
+if (! is.null(opt\$spacing)){
+    lmfit_args[['spacing']] <- as.numeric(opt\$spacing)
+}
 if (! is.null(opt\$block)){
-    lmfit_args[['block']] = sample.sheet[[opt\$block]]
+    lmfit_args[['block']] <- sample.sheet[[opt\$block]]
 }
 if (! is.null(opt\$correlation)){
-    lmfit_args[['correlation']] = opt\$correlation
+    lmfit_args[['correlation']] <- as.numeric(opt\$correlation)
+}
+if (! is.null(opt\$method)){
+    lmfit_args[['method']] <- opt\$method
 }
 
 fit <- do.call(lmFit, lmfit_args)
 
 # Contrasts bit
-contrast <- paste(paste(contrast_variable, c(opt\$target_level, opt\$reference_level), sep='.'), collapse='-')
-contrast.matrix <- makeContrasts(contrasts=contrast, levels=design)
+
+# Create the contrast string for the specified comparison
+
+treatment_target = paste0(contrast_variable, ".", opt\$target_level)
+treatment_reference = paste0(contrast_variable, ".", opt\$reference_level)
+
+if ((treatment_target %in% colnames(design)) && (treatment_reference %in% colnames(design)))  {
+    contrast_string <- paste0(treatment_target, "-", treatment_reference)
+} else if (!(treatment_target %in% colnames(design)) && (treatment_reference %in% colnames(design))) {
+    contrast_string <- paste0(treatment_reference, "-0")
+} else if ((treatment_target %in% colnames(design)) && !(treatment_reference %in% colnames(design)))  {
+    contrast_string <- paste0(treatment_target, "-0")
+} else {
+    stop(paste0(treatment_target, " and ", treatment_reference, " Not found in design array"))
+}
```

Comment:
So I've added this logic and cleaned up PR. 
What is left is to figure out the environment stuff.